components/secret-mgt/org.wso2.carbon.identity.secret.mgt.core/src/main/java/org/wso2/carbon/identity/secret/mgt/core/dao/impl/CachedBackedSecretDAO.java

```diff
@@ -0,0 +1,232 @@
+import org.wso2.carbon.identity.secret.mgt.core.constant.SecretConstants;
+import org.wso2.carbon.identity.secret.mgt.core.dao.SecretDAO;
+import org.wso2.carbon.identity.secret.mgt.core.exception.SecretManagementException;
+import org.wso2.carbon.identity.secret.mgt.core.internal.SecretManagerComponentDataHolder;
+import org.wso2.carbon.identity.secret.mgt.core.model.Secret;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.util.List;
+
+/**
+ * This is a wrapper data access object to the default data access object to provide caching functionalities.
+ */
+public class CachedBackedSecretDAO implements SecretDAO {
+
+    private static final Log log = LogFactory.getLog(CachedBackedSecretDAO.class);
+    private final SecretDAO secretDAO;
+    private final SecretByIdCache secretByIdCache;
+    private final SecretByNameCache secretByNameCache;
+
+    public CachedBackedSecretDAO(SecretDAO secretDAO) {
+
+        this.secretDAO = secretDAO;
+        this.secretByIdCache = SecretByIdCache.getInstance();
+        this.secretByNameCache = SecretByNameCache.getInstance();
+    }
+
+    @Override
+    public int getPriority() {
+
+        return 2;
+    }
+
+    @Override
+    public Secret getSecretByName(String name, int tenantId) throws SecretManagementException {
+
+        Secret secret = getSecretFromCacheByName(name, tenantId);
+        if (secret != null) {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache hit for secret by it's name. Secret name: %s, Tenant id: "
+                        + "%d", name, tenantId);
+                log.debug(message);
+            }
+            return secret;
+        } else {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache miss for secret by it's name. Secret name: %s, Tenant id: " +
+                        "%d", name, tenantId);
+                log.debug(message);
+            }
+            secret = secretDAO.getSecretByName(name, tenantId);
+            addSecretToCache(secret);
+        }
+        return secret;
+    }
+
+    @Override
+    public Secret getSecretById(String secretId, int tenantId) throws SecretManagementException {
+
+        Secret secret = getSecretFromCacheById(secretId, tenantId);
+        if (secret != null) {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache hit for secret by it's id. Secret id: %s", secretId);
+                log.debug(message);
+            }
+            return secret;
+        } else {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache miss for secret by it's id. Secret id: %s", secretId);
+                log.debug(message);
+            }
+            secret = secretDAO.getSecretById(secretId, tenantId);
+            addSecretToCache(secret);
+        }
+        return secret;
```

Comment:
Isn't it enough to return secret at this point  without returning it in the line 93 as well ?

Reply:
Addressed with commit [337cab4](https://github.com/wso2/carbon-identity-framework/pull/3770)